### PR TITLE
fix(es/codegen): Fix codegen of string literals

### PR DESCRIPTION
**Pull request by alexander-akait**
**Description:**

feat - compress strings better + improve perf for minify due to avoid using span

Also use `"` quotes by default terser and esbuild use them for minified source (maybe related to gzip)

The next steps:
1. Adding `raw` and avoid using span and source code (it will be even more faster)
2. Remove unnecessary `has_escape` and `kind`
3. Refactor and simplify code
4. Add `ascii_only` option

We should careful look at changed tests, maybe I missing something

**BREAKING CHANGE:**

No

**Related issue (if exists):**

No

### CODE REVIEW

File: crates/swc_css_parser/tests/fixture/at-rule/unknown/output.json [hunk 1]
```diff
@@ -5802,6 +5803,7 @@
       "span": {
         "start": 1664,
         "end": 1674,
+        "end": 1675,
```
Comment (kdy1) [inline crates/swc_css_parser/tests/fixture/at-rule/unknown/output.json:5806]: Seems like a wrong rebase?

Comment (alexander-akait) [reply crates/swc_css_parser/tests/fixture/at-rule/unknown/output.json:5806]: Yep, WIP on fix, my bad :smile:

File: crates/swc_ecma_codegen/src/expr.rs [hunk 2]
```diff
@@ -13,7 +13,7 @@ mod tests {
         assert_min("false", "false");
         assert_min("42", "42");
         assert_min("3.14", "3.14");
-        assert_min(r#" 'foobar' "#, r#"'foobar'"#);
+        assert_min(r#" 'foobar' "#, r#""foobar""#);
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/src/expr.rs:16]: By default we prefer `"` as do terser and esbuild, so quotes changed are expected

File: crates/swc_ecma_codegen/src/lib.rs [hunk 3]
```diff
@@ -492,42 +492,49 @@ where
...
 
-        let (single_quote, value) = match node.kind {
+        match node.kind {
             StrKind::Normal { contains_quote } => {
-                let single_quote = if contains_quote {
-                    is_single_quote(&self.cm, node.span)
-                } else {
-                    None
-                };
+                let is_invalid_use_strict = node.has_escape && &*node.value == "use strict";
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/src/lib.rs:497]: It will be improved when we will have `raw`, because we will know how developer writes it, currently we need this hack

File: crates/swc_ecma_codegen/src/lib.rs [hunk 4]
```diff
@@ -3120,45 +3127,61 @@ fn unescape_tpl_lit(s: &str, is_synthesized: bool) -> String {
...
     target: EsVersion,
     single_quote: bool,
     emit_non_ascii_as_unicode: bool,
+    pure_quotes: bool,
 ) -> String {
     let mut buf = String::with_capacity(v.len());
     let mut iter = v.chars().peekable();
 
+    let mut sq = 0;
+    let mut dq = 0;
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/src/lib.rs:3141]: We count number of quotes and get better quotes (better compressed size)

File: crates/swc_ecma_codegen/src/lib.rs [hunk 5]
```diff
@@ -3205,41 +3239,34 @@ fn escape_with_source(
...
             target,
             single_quote.unwrap_or(false),
             emit_non_ascii_as_unicode,
+            false,
         );
     }
 
-    if single_quote.is_some() && orig.len() <= 2 {
-        return escape_without_source(s, target, single_quote.unwrap_or(false), false);
-    }
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/src/lib.rs:3233]: Not sure why we need it

File: crates/swc_ecma_codegen/src/tests.rs [hunk 6]
```diff
@@ -551,40 +551,40 @@ CONTENT\r
...
+            super::get_unquoted_utf16(src, EsVersion::Es5, true, true, false),
             expected
         )
     }
 
     es2020("abcde", "abcde");
     es2020(
         "\x00\r\n\u{85}\u{2028}\u{2029};",
-        "\\x00\\r\\n\\x85\\u2028\\u2029;",
+        "\\0\\r\\n\\x85\\u2028\\u2029;",
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/src/tests.rs:579]: terser and esbuild have this output

File: crates/swc_ecma_codegen/src/tests.rs [hunk 7]
```diff
@@ -551,40 +551,40 @@ CONTENT\r
...
     );
 
     es2020("\n", "\\n");
     es2020("\t", "\\t");
 
     es2020("'string'", "\\'string\\'");
 
-    es2020("\u{0}", "\\x00");
+    es2020("\u{0}", "\\0");
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/src/tests.rs:587]: terser and esbuild have the same output too

File: crates/swc_ecma_codegen/tests/test262-min/0aeb95f62766e684.js [hunk 8]
```diff
@@ -1 +1 @@
-"\u0061"
+"a"
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/tests/test262-min/0aeb95f62766e684.js:1]: esbuild and terser have same output, less size

File: crates/swc_ecma_codegen/tests/test262-min/0b1fc7208759253b.js [hunk 9]
```diff
@@ -1 +1 @@
...
+"\0\x01\x02\x03\x04\x05\x06\x07\b\t\n\v\f\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0\xb1\xb2\xb3\xb4\xb5\xb6\xb7\xb8\xb9\xba\xbb\xbc\xbd\xbe\xbf\xc0\xc1\xc2\xc3\xc4\xc5\xc6\xc7\xc8\xc9\xca\xcb\xcc\xcd\xce\xcf\xd0\xd1\xd2\xd3\xd4\xd5\xd6\xd7\xd8\xd9\xda\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9\xfa\xfb\xfc\xfd\xfe"+a
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/tests/test262-min/0b1fc7208759253b.js:1]: Here interesting place:
esbuild:
```
{
  warnings: [],
  code: 'console.log(`\\0\u0001\u0002\u0003\u0004\u0005\u0006\\x07\\b\t\n' +
    '\\v\\f\\r\u000e\u000f\u0010\u0011\u0012\u0013\u0014\u0015\u0016\u0017\u0018\u0019\u001a\\x1B\u001c\u001d\u001e\u001f !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\\\]^_\\`abcdefghijklmnopqrstuvwxyz{|}~\\x7F\\x80\\x81\\x82\\x83\\x84\\x85\\x86\\x87\\x88\\x89\\x8A\\x8B\\x8C\\x8D\\x8E\\x8F\\x90\\x91\\x92\\x93\\x94\\x95\\x96\\x97\\x98\\x99\\x9A\\x9B\\x9C\\x9D\\x9E\\x9F\\xA0\\xA1\\xA2\\xA3\\xA4\\xA5\\xA6\\xA7\\xA8\\xA9\\xAA\\xAB\\xAC\\xAD\\xAE\\xAF\\xB0\\xB1\\xB2\\xB3\\xB4\\xB5\\xB6\\xB7\\xB8\\xB9\\xBA\\xBB\\xBC\\xBD\\xBE\\xBF\\xC0\\xC1\\xC2\\xC3\\xC4\\xC5\\xC6\\xC7\\xC8\\xC9\\xCA\\xCB\\xCC\\xCD\\xCE\\xCF\\xD0\\xD1\\xD2\\xD3\\xD4\\xD5\\xD6\\xD7\\xD8\\xD9\\xDA\\xDB\\xDC\\xDD\\xDE\\xDF\\xE0\\xE1\\xE2\\xE3\\xE4\\xE5\\xE6\\xE7\\xE8\\xE9\\xEA\\xEB\\xEC\\xED\\xEE\\xEF\\xF0\\xF1\\xF2\\xF3\\xF4\\xF5\\xF6\\xF7\\xF8\\xF9\\xFA\\xFB\\xFC\\xFD\\xFE`);\n',
  map: ''
}
```

terser:
```
{
    code: 'console.log("\\0\u0001\u0002\u0003\u0004\u0005\u0006\u0007\\b\\t\\n\\v\\f\\r\u000e\u000f\u0010\u0011\u0012\u0013\u0014\u0015\u0016\u0017\u0018\u0019\u001a\u001b\u001c\u001d\u001e\u001f !\\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþ");'
  }
```

Comment (alexander-akait) [reply crates/swc_ecma_codegen/tests/test262-min/0b1fc7208759253b.js:1]: terser and esbuild prefer using `\u` over `\x`, I will investigate

Comment (alexander-akait) [reply crates/swc_ecma_codegen/tests/test262-min/0b1fc7208759253b.js:1]: Other changes is almost the same, because they just contain characters from this test

File: crates/swc_ecma_codegen/src/lib.rs [hunk 10]
```diff
@@ -3170,6 +3193,9 @@ fn escape_without_source(
             '\u{2029}' => {
                 buf.push_str("\\u2029");
             }
+            '\u{FEFF}' => {
```
Comment (kdy1) [inline crates/swc_ecma_codegen/src/lib.rs:3196]: Why is this case required?

File: crates/swc_ecma_codegen/tests/test262-min/e9682c37a1a959e1.js [hunk 11]
```diff
@@ -1 +1 @@
-"\u{20BB7}\u{91CE}\u{5BB6}"
```
Comment (kdy1) [inline crates/swc_ecma_codegen/tests/test262-min/e9682c37a1a959e1.js:1]: Is this change intentional?

**Review by kdy1 — COMMENTED** [review]
Thanks!
I have some questions

**Comment by alexander-akait** [comment]
Interesting - https://v8.dev/blog/scanner, shortly - we should avoid unicode for identifier (and not only) too, yes, file is less, but parser is slow, no golden way

File: crates/swc_ecma_codegen/src/lib.rs (hunk 10 above)
Comment (alexander-akait) [reply crates/swc_ecma_codegen/src/lib.rs:3196]: Yes, otherwise we will print invalid character:
esbuild - https://github.com/evanw/esbuild/blob/master/internal/js_printer/js_printer.go#L274
terser - https://github.com/terser/terser/blob/master/lib/output.js#L330

File: crates/swc_ecma_codegen/tests/test262-min/e9682c37a1a959e1.js (hunk 11 above)
Comment (alexander-akait) [reply crates/swc_ecma_codegen/tests/test262-min/e9682c37a1a959e1.js:1]: Related https://github.com/swc-project/swc/pull/4037/#issuecomment-1068468119, I think we should avoid it, unicode parsing is slow

File: crates/swc_ecma_codegen/tests/test262-min/664b0da1dd015106.js [hunk 12]
```diff
@@ -1 +1 @@
-a["b"]="c";a["if"]="if";a["*"]="d";a["\u0EB3"]="e";a[""]="f"
```
Comment (alexander-akait) [inline crates/swc_ecma_codegen/tests/test262-min/664b0da1dd015106.js:1]: Here the same, esbuild prefer to print `\u0EB3`, terser - real unicode character

File: crates/swc/tests/fixture/issue-1259/output/index.js [hunk 13]
```diff
@@ -1,7 +1,7 @@
 var swcHelpers = require("@swc/helpers");
 function _templateObject() {
     var data = swcHelpers.taggedTemplateLiteral([
-        "'#ERROR'"
+        "#ERROR"
```
Comment (kdy1) [inline crates/swc/tests/fixture/issue-1259/output/index.js:4]: This seems like a wrong change.

**Review by kdy1 — CHANGES_REQUESTED** [review]
I'm reviewing other files, too.

File: crates/swc/tests/tsc-references/parserindenter_es5.1.normal.js [hunk 14]
```diff
@@ -380,7 +380,7 @@ var Formatting;
             return null;
         };
         _proto.GetIndentationDelta = function GetIndentationDelta(tokenStartPosition, childTokenStartPosition /*?*/ ) {
-            Debug.Assert(childTokenStartPosition !== undefined, "Error: caller must pass 'null' for undefined position");
+            Debug.Assert(childTokenStartPosition !== undefined, "Error: caller must pass null for undefined position");
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/parserindenter_es5.1.normal.js:383]: This seem like a wrong change

File: crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js [hunk 15]
```diff
@@ -8,133 +8,133 @@
...
 }
 var Ж = 1;
 if (Ж !== 1) {
-    $ERROR('#Ж');
+    $ERROR('#\u0416');
 }
 var З = 1;
 if (З !== 1) {
-    $ERROR('#З');
+    $ERROR('#\u0417');
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js:39]: Is it fine to emit this for `es5`?

File: crates/swc/tests/tsc-references/jsDeclarationsReferenceToClassInstanceCrossFile_es5.1.normal.js [hunk 16]
```diff
@@ -8,7 +8,7 @@ import * as swcHelpers from "@swc/helpers";
 var Rectangle = function Rectangle() {
     "use strict";
     swcHelpers.classCallCheck(this, Rectangle);
-    console.log("I'm a rectangle!");
+    console.log("Im a rectangle!");
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/jsDeclarationsReferenceToClassInstanceCrossFile_es5.1.normal.js:11]: I think this is also an issue

**Review by kdy1 — CHANGES_REQUESTED** [review]
I'm reviewing other files, too.

Btw, the previous code is written in that way to preserve input as much as possible.
I'm not sure if it's fine to apply this patch without some complement

File: crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js (hunk 15 above)
Comment (alexander-akait) [reply crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js:39]: Yes, it is not real `3`, it is unicode character, just try `'З' === '3'` in console

Comment (kdy1) [reply crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js:39]: I mean that I'm not sure if `\u0417`  is fine for es5

Comment (alexander-akait) [reply crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js:39]: It is fine, es5 support `\uXXXX`, es5 not supports `\u{XXXXXX}`, spec https://262.ecma-international.org/5.1/#sec-6

Comment (kdy1) [reply crates/swc/tests/tsc-references/parserS7.6_A4.2_T1_es5.1.normal.js:39]: Oh I see, thanks!

File: crates/swc_ecma_codegen/tests/test262-min/cb625ce2970fe52a.js [hunk 17]
```diff
@@ -1 +1 @@
-(function(){'use\x20strict';with(a);})
+(function(){'use strict';with(a);})
```
Comment (kdy1) [inline crates/swc_ecma_codegen/tests/test262-min/cb625ce2970fe52a.js:1]: I think this is a bomb, although I don't expect someone to write code like `'use\x20strict'` unless they are testing...

File: crates/swc_ecma_codegen/tests/test262-min/ac73bc36bbc48890.js [hunk 18]
```diff
@@ -1 +1 @@
-"use\x20strict";with(a)b=c
+"use strict";with(a)b=c
```
Comment (kdy1) [inline crates/swc_ecma_codegen/tests/test262-min/ac73bc36bbc48890.js:1]: Same here

**Review by kdy1 — COMMENTED** [review]
I found some bugs related to escapes in `use strict` maybe transform is setting `has_escape` to false?

File: crates/swc_ecma_codegen/tests/test262-min/855b8dea36c841ed.js [hunk 19]
```diff
@@ -1 +1 @@
-(function(){'use\x20strict';with(a);}())
+(function(){'use strict';with(a);}())
```
Comment (kdy1) [inline crates/swc_ecma_codegen/tests/test262-min/855b8dea36c841ed.js:1]: One more

**Review by kdy1 — CHANGES_REQUESTED** [review]
I found some more issues

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_codegen
